cpp/src/arrow/util/checked-cast-test.cc

```diff
@@ -0,0 +1,40 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <type_traits>
+
+#include <gtest/gtest.h>
+
+#include "arrow/util/checked_cast.h"
+
+namespace arrow {
+
+TEST(CheckedCast, TestInvalidSubclassCast) {
+  class Foo {
+   public:
+    virtual ~Foo() = default;
+  };
+  class Bar {};
+
+  static_assert(std::is_polymorphic<Foo>::value, "Foo is not polymorphic");
+
+  Foo foo;
+  Foo* fooptr = &foo;
+  ASSERT_EQ(nullptr, checked_cast<Bar*>(fooptr));
```

Comment:
Won't this test fail in non-debug mode?

Reply:
(also I think the test would be more interesting if there was a inheritance relationship between Foo and Bar, but perhaps it doesn't change anything).

Reply:
Yes, actually it will fail at compile time. Let me fix